syntax = "proto3";

// Interface exported by the server
service ClientManager {
  rpc getClients(GetClientsMsg) returns (ClientsDTO) {}
  rpc stopClients(StopClientsMsg) returns (ClientsDTO) {}
  rpc startClients(StartClientsMsg) returns (ClientsDTO) {}
  rpc startProducer(StartProducerMsg) returns (ClientsDTO) {}
  rpc stopProducer(StopProducerMsg) returns (ClientsDTO) {}
}

// Message that the client sends to stop the clients processes
message StopClientsMsg {}

// Message that the client sends to start the clients processes
message StartClientsMsg {
  int32 time_step_len_seconds = 1;
  WorkflowsConfigDTO workflows_config = 2;
  repeated ClientDTO clients = 3;
}

// ClientDTO
message ClientDTO {
  int32 id = 1;
  repeated float workflow_distribution = 2;
  ConstantArrivalConfigDTO constant_arrival_config = 3;
  SineArrivalConfigDTO sine_arrival_config = 4;
  SpikingArrivalConfigDTO spiking_arrival_config = 5;
  PieceWiseConstantArrivalConfigDTO piece_wise_constant_arrival_config = 6;
  EPTMPArrivalConfigDTO eptmp_arrival_config = 7;
  float mu = 9;
  bool exponential_service_time = 10;
  int32 arrival_type = 11;
}

// Message that the client sends to extract the ClientsDTO
message GetClientsMsg {}

// Message that the client sends to start the producer thread
message StartProducerMsg {
  string ip = 1;
  int32 port = 2;
  int32 time_step_len_seconds = 3;
}

// Message that the client sends to start the producer thread
message StopProducerMsg {}

// ProbabilityDistributionDTO
message ProbabilityDistributionDTO {
  repeated float probabilities = 1;
}

// TransitionMatrixDTO
message TransitionMatrixDTO {
  repeated ProbabilityDistributionDTO rows = 1;
}

// WorkflowMarkovChainDTO
message WorkflowMarkovChainDTO {
  TransitionMatrixDTO transition_matrix = 1;
  int32 initial_state = 2;
  int32 id = 3;
}

// WorkflowServiceDTO
message WorkflowServiceDTO {
  int32 id = 1;
  repeated string ips = 2;
  repeated NodeCommandsDTO commands = 3;
}

// NodeCommandsDTO
message NodeCommandsDTO {
  repeated string commands = 1;
}

// WorkflowServiceDTO
message WorkflowsConfigDTO {
  repeated WorkflowMarkovChainDTO workflow_markov_chains = 1;
  repeated WorkflowServiceDTO workflow_services = 2;
}

// Message that the server returns when requested by the client
message ClientsDTO {
  int32 num_clients = 1;
  bool client_process_active = 2;
  bool producer_active = 3;
  int32 clients_time_step_len_seconds = 4;
  int32 producer_time_step_len_seconds = 5;
}

// ConstantArrivalConfigDTO
message ConstantArrivalConfigDTO {
  float lamb = 1;
}

// SineArrivalConfigDTO
message SineArrivalConfigDTO {
  float lamb = 1;
  float time_scaling_factor = 2;
  float period_scaling_factor = 3;
}

// SpikingArrivalConfigDTO
message SpikingArrivalConfigDTO {
  repeated float exponents = 1;
  repeated float factors = 2;
}

// PieceWiseConstantArrivalConfigDTO
message PieceWiseConstantArrivalConfigDTO {
  repeated float breakvalues = 1;
  repeated int32 breakpoints = 2;
}

// EPTMPArrivalConfigDTO
message EPTMPArrivalConfigDTO {
  repeated float thetas = 1;
  repeated float gammas = 2;
  repeated float phis = 3;
  repeated float omegas = 4;
}